typescript/src/lib/utils/ledger.ts

```diff
@@ -0,0 +1,177 @@
+import { ethers, Signer } from "ethers"
+import {
+  Account,
+  WalletAPIClient,
+  WindowMessageTransport,
+} from "@ledgerhq/wallet-api-client"
+import BigNumber from "bignumber.js"
+import { AddressZero } from "@ethersproject/constants"
+import { Deferrable } from "@ethersproject/properties"
+import { Hex } from "./hex"
+class AccountNotFoundError extends Error {
+  constructor() {
+    super(
+      "Account not found. Please use `requestAccount` method or set the signer account with `setAccount` method."
+    )
+  }
+}
+
+export class LedgerLiveAppEthereumSigner extends Signer {
+  private _walletApiClient: WalletAPIClient
+  private _windowMessageTransport: WindowMessageTransport
+  private _account: Account | undefined
+
+  constructor(
+    provider: ethers.providers.Provider,
+    windowMessageTransport?: WindowMessageTransport,
+    walletApiClient?: WalletAPIClient
+  ) {
+    super()
+    ethers.utils.defineReadOnly(this, "provider", provider || null)
```

Comment:
ddbb7e0e45587c6b60911e4d9ffaaa2abee2f036

Reply:
Good but let's leave the original `ethers.utils.defineReadOnly(this, "provider", provider || null)` initializer. This way `provider` has only one possible value (`null`) when it is not passed in the constructor. Ethers does the same in the `VoidSigner`.

Reply:
> Good but let's leave the original `ethers.utils.defineReadOnly(this, "provider", provider || null)` initializer.

Actually it does not work in ethers instance of the Signer 🤔 The only possible types are `Provider` or `undefined`. Once I've change dthe `provider` argument to optional with `?`, the error occured for this line:

```
Argument of type 'Provider | null' is not assignable to parameter of type 'this["provider"]'.
  'this["provider"]' could be instantiated with an arbitrary type which could be unrelated to 'Provider | null'.
```

so that's why I'm setting only a provider there now, which is a type of `Provider | undefined`.

Reply:
Got it!